File: src/_01_CreationalPattern/_01_05_Prototype/GithubIssue.js
import GithubRepository from "./GithubRepository.js";

export default class GithubIssue {
  /** @type { number } */
  #id;

  /**
   * 이슈명
   * @type { string }
   */
  #title;

  /**
   * 저장소
   * @type { GithubRepository }
   */
  #repository;

  /** @param { GithubRepository } repository */
  constructor(repository) {
    this.#repository = repository;
  }

  /** @returns { number } */
  getId() {
    return this.#id;
  }
  /** @param { number } id */
  setId(id) {
    this.#id = id;
  }

  /** @returns { string } */
  getTitle() {
    return this.#title;
  }
  /** @param { string } title */
  setTitle(title) {
    this.#title = title;
  }

  /** @returns { GithubRepository } */
  getRepository() {
    return this.#repository;
  }

  /** @returns { string } */
  getUrl() {
    return `
      URL: https://github.com/${this.#repository.getUser()}/${this.#repository.getName()}/${this.#id}
    `;
  }

  /**
   * Prototype 패턴 적용 메서드 (깊은복사로 구현한 예)
   * @returns { GithubIssue }
   */
  cloneDeep() {
    const originRepository = this.#repository;

    const clonedRepository = new GithubRepository();
    clonedRepository.setUser(originRepository.getUser());
    clonedRepository.setName(originRepository.getName());

    const issue = new GithubIssue(clonedRepository);
    issue.setId(this.#id);
    issue.setTitle(this.#title);

    return issue;
  }
}

File: src/_01_CreationalPattern/_01_05_Prototype/java/before/GithubIssue.java
package _01_CreationalPattern._01_05_Prototype.java.before;

public class GithubIssue {
  
  private int id;
  private String title;
  private GithubRepository repository;

  public GithubIssue(GithubRepository repository) {
    this.repository = repository;
  }

  // id
  public int getId() {
    return id;
  }
  public void setId(int id) {
    this.id = id;
  }

  // title
  public String getTitle() {
    return title;
  }
  public void setTitle(String title) {
    this.title = title;
  }

  // repository
  public GithubRepository getRepository() {
    return repository;
  }

  public String getUrl() {
    return String.format(
      "https://github.com/%s/%s/issues/%d",
      repository.getUser(),
      repository.getName(),
      id
    );
  }
  
}
